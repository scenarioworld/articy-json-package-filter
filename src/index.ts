import { LoaderDefinition } from 'webpack';
import { ArticyData } from 'articy-js';
import { validate } from 'schema-utils';
import { Schema } from 'schema-utils/declarations/validate';

const schema: Schema = {
  type: 'object',
  properties: {
    include: {
      type: 'array',
      items: {
        anyOf: [{ instanceof: 'RegExp' }, { type: 'string' }],
      },
    },
  },
};

export interface ArticyJsonPackageFilterOptions {
  include: (RegExp | string)[];
}

// Our asyncronous loader
const loader: LoaderDefinition<ArticyJsonPackageFilterOptions> = function(
  source
) {
  const options = this.getOptions();

  validate(schema, options, {
    name: 'Articy JSON Package Filter',
    baseDataPath: 'options',
  });

  // Load source as JSON
  const parsedData = JSON.parse(source) as ArticyData;

  // Get the names of each package
  const packageIds = parsedData.Packages.map(p => p.Name);

  // For each package, remove if its name doesn't match any of the include regex
  for (const packageId of packageIds) {
    if (
      !options.include.some(r =>
        typeof r === 'string' ? packageId.indexOf(r) !== -1 : r.test(packageId)
      )
    ) {
      parsedData.Packages.splice(
        parsedData.Packages.findIndex(p => p.Name === packageId),
        1
      );
    }
  }

  // Re-export
  return `${JSON.stringify(parsedData)}`;
};

export default loader;
