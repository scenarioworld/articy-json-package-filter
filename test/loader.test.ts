import { ArticyData } from 'articy-js';
import webpack from 'webpack';
import { ArticyJsonPackageFilterOptions } from '../src';
import compiler from './compiler';

async function RunTest(
  options: ArticyJsonPackageFilterOptions
): Promise<ArticyData> {
  // Compile and collect output
  const stats = await compiler('data.articy.json', options);
  let output = stats.toJson({ source: true }).modules?.[0].source!;

  // Make sure we actually get output
  expect(output).toBeDefined();

  // Convert output to a string
  if (output instanceof Buffer) {
    output = output.toString();
  }

  // Parse to JSON
  return JSON.parse(output) as ArticyData;
}

test('No include means no packages', async () => {
  const result = await RunTest({
    include: [],
  });

  expect(result.Packages).toHaveLength(0);
});

test('Only include game packages', async () => {
  const result = await RunTest({
    include: [/^game\./],
  });

  expect(result.Packages).toHaveLength(3);
  expect(result.Packages.every(p => p.Name.startsWith('game.'))).toBeTruthy();
});

test('Only include sandbox packages', async () => {
  const result = await RunTest({
    include: [/^sandbox\./],
  });

  expect(result.Packages).toHaveLength(2);
  expect(
    result.Packages.every(p => p.Name.startsWith('sandbox.'))
  ).toBeTruthy();
});

test('Include episode one and sandbox', async () => {
  const result = await RunTest({
    include: [/^sandbox\./, /ep1/],
  });

  expect(result.Packages).toHaveLength(3);
  expect(
    result.Packages.every(
      p => p.Name.startsWith('sandbox.') || p.Name.indexOf('ep1') !== -1
    )
  ).toBeTruthy();
});

test('Strings as arguments', async () => {
  const result = await RunTest({
    include: ['sandbox'],
  });

  expect(result.Packages).toHaveLength(2);
  expect(
    result.Packages.every(p => p.Name.startsWith('sandbox.'))
  ).toBeTruthy();
});

test('Bad input causes failure', async () => {
  try {
    await RunTest({
      //@ts-ignore
      include: [4],
    });
    fail('compiler should fail');
  } catch (error) {
    if (Array.isArray(error)) {
      const errors = error as webpack.StatsError[];
      expect(errors[0]).toBeDefined();
      const match = errors[0].message.match(/Error:.*$/gm);
      expect(match?.[0]).toMatchSnapshot();
    } else {
      fail('Expected an array of webpack errors.');
    }
  }
});
