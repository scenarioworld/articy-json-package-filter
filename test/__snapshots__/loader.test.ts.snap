// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bad input causes failure 1`] = `"Error: Invalid options object. Articy JSON Package Filter has been initialized using an options object that does not match the API schema."`;
